### PR TITLE
Add progress callback to TS client

### PR DESCRIPTION
**Pull request by zachmullen**
This adds progress reporting support on top of #161 

The API for progress events I'm proposing here is extremely explicit. It makes no attempt to fake progress during init and finalize, rather pushing that responsibility out to downstreams if they desire that behavior. Other clients (like the one I'm writing) can instead opt to show indeterminate progress, or even a helpful message about what's going on during the init and finalization periods.

### CODE REVIEW

**Comment by zachmullen** [comment]
This is now ready for review @brianhelba

File: django-s3-file-field-client/src/client.ts
Comment (brianhelba) [inline django-s3-file-field-client/src/client.ts:39]: Do you think it'd be helpful to ever know when there's no callback attached?

If so, we should make `onProgress` default to `undefined`.

I'm not sure if this is something we'd ever care about though. It's just a question, not a suggestion.

File: django-s3-file-field-client/src/client.ts
Comment (brianhelba) [inline django-s3-file-field-client/src/client.ts:65]: Could you rename `index` to be something more descriptive now, like `file_offset` (or similar)?

File: django-s3-file-field-client/src/client.ts
Comment (brianhelba) [inline django-s3-file-field-client/src/client.ts:125]: This should start at the beginning of `completeUpload`, since that's part of the asynchronous "overhead" work.

I think you could move this invocation to `uploadFile`, since it doesn't really need to be inside the `protected` helper.

File: django-s3-file-field-client/src/client.ts
Comment (brianhelba) [inline django-s3-file-field-client/src/client.ts:128]: Once this completes, it'd be nice to set the progress to some terminal state. I think this can also be set at the end of `uploadFile`.

File: django-s3-file-field-client/src/client.ts
Comment (brianhelba) [inline django-s3-file-field-client/src/client.ts:29]:
```suggestion
  uploaded?: number;
```

Seems like a more sensible name.

File: django-s3-file-field-client/src/client.ts
Comment (brianhelba) [inline django-s3-file-field-client/src/client.ts:31]:
```suggestion
  readonly loaded?: number;
  readonly total?: number;
  readonly state: 'initializing' | 'sending' | 'finalizing';
```

Comment (brianhelba) [inline django-s3-file-field-client/src/client.ts:31]: Could this value be an `enum`?

**Review by brianhelba — COMMENTED** [review]
I like this approach, and it's certainly much more straightforward now that parts are uploaded strictly serially.

I also agree that it's better to keep the "overhead" portions indeterminate, as opposed to #147.

See some small suggestions and questions inline, but after that's addressed, I approve.

File: django-s3-file-field-client/src/client.ts
Comment (zachmullen) [reply django-s3-file-field-client/src/client.ts:31]: I was following the pattern from line 25 above. For the sake of consistency, we should use the same pattern in both places, whichever we choose. (I leave it up to you to pick which one is best.)

File: django-s3-file-field-client/src/client.ts
Comment (zachmullen) [reply django-s3-file-field-client/src/client.ts:39]: I think this default value is an internal detail, so if it ever becomes a requirement for us to know whether it was passed in or not, we can change it without any issue. Absent that, using a non-optional function type greatly simplifies the implementation, so I'd prefer to leave it.

File: django-s3-file-field-client/src/client.ts
Comment (zachmullen) [reply django-s3-file-field-client/src/client.ts:128]: Maybe so, although at that point it's functionally equivalent to just awaiting the Promise resolution, so it doesn't add practical value. If you feel strongly though, I'll add it.

Comment (brianhelba) [reply django-s3-file-field-client/src/client.ts:126]: I think it can make things simpler for implementors, depending on how they structure things. I think it's worth adding.

File: django-s3-file-field-client/src/client.ts
Comment (brianhelba) [reply django-s3-file-field-client/src/client.ts:39]: Agreed.

File: django-s3-file-field-client/src/client.ts
Comment (brianhelba) [reply django-s3-file-field-client/src/client.ts:126]: Less assembly required is usually better for downstream usability.

File: django-s3-file-field-client/src/client.ts
Comment (brianhelba) [reply django-s3-file-field-client/src/client.ts]: I see. For TypeScript, I think `enum` is the right choice in both places.

**Comment by zachmullen** [comment]
@brianhelba This is ready for final review. Note that I also changed `UploadResult.state` to be an enum for consistency. Please evaluate whether that will have any breaking impact outside of this library.

**Review by brianhelba — APPROVED** [review]

**Comment by brianhelba** [comment]
There are very few downstreams using this library directly at the moment, though I will include a release note about the change.

I don't consider any part of S3FF to be stable currently, though I'd like to avoid changing public interfaces (which this is) unless there's a good reason (which I also think this is).

File: django-s3-file-field-client/src/client.ts
Comment (brianhelba) [inline django-s3-file-field-client/src/client.ts:52]: @zachmullen This should be `protected`.

I'll push a fix as part of another PR.

**Pull request merged**
